### DISCUSSION TITLE
【HELP!】Is there a way to extract client IP? 【Next13 - APP dir】

### DISCUSSION THREAD

**Discussion by jimjimliu** [vercel/community#4634] [Help]
### Summary

【general】
version: v13.4.8 (app dir)
request method: route handlers
backend: springboot

【summary】
Does anyone know how I can get client IP addr in middleware.js OR from springboot backend?

【description】
After Next introduced [route handlers], a request chain is like this [client -> route handler -> springboot], and technically route handler is on server side. Therefore, I've tried many ways and it seems I cannot get the actual client IP address from a request made to my springboot backend. The IP extracted from request in this way is the IP of my own server, instead of the client IP.

SO, how can i get a client IP, in middleware.js or in springboot. Anywhere i can get the actual client IP will do.

【methods I’'ve tried】
1, abandon route handlers, use axios to make requests. 
2, before a request, make a separate request to some vendors' open API to get client's ip. and manually put the ip into header with whatever name one likes.
3, request to some vendors' open API to get the client ip, and write it into cookies. Then extract client ip from cookies in springboot

Those were some methods I have tried, but none of them seems secure and good enough.  

### Steps to Reproduce

.

**Comment by nathanjhood** [comment]
Yes. [My current middleware.ts](https://github.com/StoneyDSP/stoneydsp.com/blob/production/middleware.ts) does this:

```.tsx
import { NextResponse } from 'next/server'

import isbot from 'isbot'

import type { NextRequest } from 'next/server'

export async function middleware(req: NextRequest) {

  const res = NextResponse.next()

  // Extract visitor info. 
  const country = (req.geo && req.geo.country) || 'Earth'
  const city = (req.geo && req.geo.city) || 'Nowhere'
  const region = (req.geo && req.geo.region) || 'Somewhere'
  const ip = (req.ip) || 'Visitor'
  const agent = (req.headers.get('user-agent')) || 'Agent Unknown'

  if (isbot(req.headers.get('user-agent'))) {
    console.log(`Bot ${ip} crawling from ${city}, ${region}, ${country} with ${agent}`)
  } else {
    console.log(`User ${ip} visiting from ${city}, ${region}, ${country} with ${agent} `)
  }
  return res
}

export const config = {
  matcher: [
    /*
     * Match all request paths except for the ones starting with:
     * - api (API routes)
     * - _next/static (static files)
     * - _next/image (image optimization files)
     * - favicon.ico (favicon file)
     */
    {
      source: '/((?!api|_next/static|_next/image|favicon.ico).*)',
      missing: [
        { type: 'header', key: 'next-router-prefetch' },
        { type: 'header', key: 'purpose', value: 'prefetch' },
      ],
    },
  ],
}
```

Sample output as seen in vercel domain logs:

```
Bot 199.45.155.33 crawling from Nowhere, Somewhere, US with Mozilla/5.0 (compatible; CensysInspect/1.1; +https://about.censys.io/)
```

It uses the 'isbot' npm package as a runtime dependency. This may be useful for doing things like filtering your logs. You can skip the 'isbot' dependency, along with the import and the 'if()' statement, if not needed.

I advise to make it clear in your terms of service that you are logging visitor's IP addresses and user agents, and such forth.
